### PR TITLE
streamingccl: add ingestion job framework

### LINKED ISSUE #57399: streaming: create stream ingestion job
A new type of job is needed to consume a partitioned stream.

The job should: 
- Using a streaming client, fetch the stream partitions from a "stream address".
- Create a DistSQL flow of stream partition ingestion processors (introduced in https://github.com/cockroachdb/cockroach/issues/57398)

The initial version will explicitly not require coordination of resolved timestamps across processors to keep track of the low-watermark.

### PR DESCRIPTION
**Pull request by adityamaru**
This change introduces a new StreamIngestionJob. It does not do much
more than laying out the general outline of the job, which is very
similar to other bulk jobs such as changefeed, backup etc.

More precisely:
- Introduces StreamIngestionDetails job details proto
- Hooks up the dependency to a mock stream client
- Introduces a StreamIngestionProcessorSpec
- Sets up a simple DistSQL flow which round-robin assigns the partitions
  to the processors.

Most notable TODOs in job land which will be addressed in follow up PRs:
- StreamIngestionPlanHook to create this job. It Will involve figuring out
  SQL syntax.
- Introducing a ts watermark in both the job and processors. This watermark will represent the lowest resolved ts which all processors
have ingested till. Iron out semantics on job start and resumption.
- Introducing a StreamIngestionFrontier processor which will slurp the
  results from the StreamIngestionProcessors, and use them to keep track
of the minimum resolved ts across all processors.

Fixes: #57399

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Rough WIP to ensure that the PR doesn't become hard to review. I have left some TODOs in the code and would be happy to receive any comments on them.

File: pkg/jobs/jobspb/jobs.proto [hunk 1]
```diff
@@ -77,6 +77,11 @@ message EncryptionInfo {
   map<string, bytes> encryptedDataKeyByKMSMasterKeyID = 3;
 }
 
+message StreamIngestionDetails {
+  // StreamAddress is the location of the stream which the ingestion job will
+  // read from.
+  string stream_address = 1;
```
Comment (pbardea) [inline pkg/jobs/jobspb/jobs.proto:83]: `casttype` to a `streamclient.StreamAddress`?

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 2]
```diff
@@ -134,6 +134,10 @@ message ReadImportDataSpec {
   // NEXTID: 16
 }
 
+message StreamIngestionDataSpec {
+  map<int32,string> partition_address = 1;
```
Comment (pbardea) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:138]: I think we can also add `[(gogoproto.castvalue) = "github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.PartitionAddress"];` on this line to get this map to be typed a bit stricter. I also think we can `castkey` to a `PartitionID` type that I can add to the `streamclient` package.

**Review by pbardea — APPROVED** [review]
This looks good to me, I left a few small casting nits that rely on changes to the stream client PR. I'll ping here when that is updated for a rebase.

File: pkg/jobs/jobspb/jobs.proto (hunk 1 above)
Comment (adityamaru) [reply pkg/jobs/jobspb/jobs.proto]: done.

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 2 above)
Comment (adityamaru) [reply pkg/sql/execinfrapb/processors_bulk_io.proto]: changed it to a repeated field as that is what the ingestion processor PR changed it to.

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58373)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @pbardea)

___
*[pkg/ccl/streamingccl/streamclient/mock_stream_client.go, line 19 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58373#-MQSe3Wn3w_gJQ80ibO8:-MQSe3Wn3w_gJQ80ibO9:beik9mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/da56c2205296e648ea5c8c43fa3766e4c120a695/pkg/ccl/streamingccl/streamclient/mock_stream_client.go#L19)):*
> ```Go
> 
> // NewMockStreamClient returns a new mock stream client.
> func NewMockStreamClient() *MockStreamClient {
> ```

What do you think of renaming this to something like just `client` and making it the real stream client that just isn't yet implemented? (I also don't think it needs to be exported?)
Then this can just return the interface.

**Review by pbardea — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!

bors r=pbardea

**Comment by adityamaru** [comment]
looks like a flake in `acceptance/gossip/peerings`.

bors r=pbardea

**Pull request merged**